types/wlr_surface.c

```diff
@@ -90,23 +93,64 @@ static void surface_set_input_region(struct wl_client *client,
 	}
 }
 
+static void wlr_surface_to_buffer_region(struct wlr_surface *surface,
+		pixman_region32_t *surface_region, pixman_region32_t *buffer_region) {
+	pixman_box32_t *src_rects, *dest_rects;
+	int nrects, i;
+	int scale = surface->current.scale;
+
+	src_rects = pixman_region32_rectangles(surface_region, &nrects);
+	dest_rects = malloc(nrects * sizeof(*dest_rects));
+	if (!dest_rects)
+		return;
+
+	for (i = 0; i < nrects; i++) {
+		dest_rects[i].x1 = src_rects[i].x1 * scale;
+		dest_rects[i].y1 = src_rects[i].y1 * scale;
+		dest_rects[i].x2 = src_rects[i].x2 * scale;
+		dest_rects[i].y2 = src_rects[i].y2 * scale;
+	}
+
+	pixman_region32_fini(buffer_region);
+	pixman_region32_init_rects(buffer_region, dest_rects, nrects);
+	free(dest_rects);
+}
+
 static void surface_commit(struct wl_client *client,
 		struct wl_resource *resource) {
 	struct wlr_surface *surface = wl_resource_get_user_data(resource);
+	surface->current.scale = surface->pending.scale;
 
 	if ((surface->pending.invalid & WLR_SURFACE_INVALID_BUFFER)) {
 		surface->current.buffer = surface->pending.buffer;
 	}
 	if ((surface->pending.invalid & WLR_SURFACE_INVALID_SURFACE_DAMAGE)) {
-		// TODO: Sort out buffer damage too
 		pixman_region32_union(&surface->current.surface_damage,
 				&surface->current.surface_damage,
 				&surface->pending.surface_damage);
+
+		pixman_region32_union(&surface->current.buffer_damage,
+			&surface->current.buffer_damage,
+			&surface->pending.buffer_damage);
+
+		pixman_region32_t buffer_damage;
+		wlr_surface_to_buffer_region(surface, &surface->current.surface_damage,
+			&buffer_damage);
+		pixman_region32_union(&surface->current.buffer_damage,
+			&surface->current.buffer_damage, &buffer_damage);
+
+		struct wl_shm_buffer *buffer = wl_shm_buffer_get(surface->current.buffer);
+		pixman_region32_intersect_rect(&surface->current.buffer_damage,
+			&surface->current.buffer_damage, 0, 0,
+			wl_shm_buffer_get_width(buffer),
+			wl_shm_buffer_get_height(buffer));
+
 		// TODO: Surface sizing is complicated
 		//pixman_region32_intersect_rect(&surface->current.surface_damage,
 		//		&surface->current.surface_damage,
 		//		0, 0, surface->width, surface->height);
```

Comment:
This commented out code is just to ensure that the damage falls within the buffer size, so if you're already handling it (seems you are, but incorrectly), then this can just be removed.